File: cpp/graphs/shortestpaths/dijkstra.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int, int> edge;
typedef pair<int, int> item;

// https://cp-algorithms.com/graph/dijkstra_sparse.html

// O(E*log(V)) time and O(E) memory
tuple<vector<int>, vector<int>> dijkstra_heap(const vector<vector<edge>> &g, int s) {
    size_t n = g.size();
    vector<int> prio(n, numeric_limits<int>::max());
    vector<int> pred(n, -1);
    priority_queue<item, vector<item>, greater<>> q;
    q.emplace(prio[s] = 0, s);

    while (!q.empty()) {
        auto [d, u] = q.top();
        q.pop();

        if (d != prio[u])
            continue;

        for (auto [v, len] : g[u]) {
            int nprio = prio[u] + len;
            if (prio[v] > nprio) {
                prio[v] = nprio;
                pred[v] = u;
                q.emplace(nprio, v);
            }
        }
    }

    return {prio, pred};
}

// O(E*log(V)) time and O(V) memory
tuple<vector<int>, vector<int>> dijkstra_set(const vector<vector<edge>> &g, int s) {
    size_t n = g.size();
    vector<int> prio(n, numeric_limits<int>::max());
    vector<int> pred(n, -1);
    set<item> q;
    q.emplace(prio[s] = 0, s);

    while (!q.empty()) {
        int u = q.begin()->second;
        q.erase(q.begin());

        for (auto [v, len] : g[u]) {
            int nprio = prio[u] + len;
            if (prio[v] > nprio) {
                q.erase({prio[v], v});
                prio[v] = nprio;
                pred[v] = u;
                q.emplace(prio[v], v);
            }
        }
    }

    return {prio, pred};
}

int main() {
    vector<vector<edge>> g(3);
    g[0].emplace_back(1, 10);
    g[1].emplace_back(2, -5);
    g[0].emplace_back(2, 8);

    auto [prio1, pred1] = dijkstra_heap(g, 0);
    auto [prio2, pred2] = dijkstra_set(g, 0);

    for (int x : prio1)
        cout << x << " ";
    cout << endl;

    for (int x : prio2)
        cout << x << " ";
    cout << endl;
}

File: rust/graphs/shortestpaths/dijkstra.rs
use std::collections::BinaryHeap;

// https://cp-algorithms.com/graph/dijkstra_sparse.html
// O(E*log(V)) time and O(E) memory
pub fn dijkstra_heap(graph: &Vec<Vec<(usize, i32)>>, s: usize) -> (Vec<i32>, Vec<usize>) {
    let n = graph.len();
    let mut prio = vec![i32::MAX; n];
    let mut pred = vec![usize::MAX; n];
    let mut q = BinaryHeap::<(i32, usize)>::new();
    prio[s] = 0;
    q.push((0, s));
    while let Some((d, u)) = q.pop() {
        if -d != prio[u] { continue; }
        for (v, len) in &graph[u] {
            let nprio = prio[u] + len;
            if prio[*v] > nprio {
                prio[*v] = nprio;
                pred[*v] = u;
                q.push((-nprio, *v));
            }
        }
    }
    return (prio, pred);
}

fn main() {
    let g = vec![vec![(1, 10), (2, 8)], vec![(2, -5)], vec![]];
    let (prio, pred) = dijkstra_heap(&g, 0);
    assert_eq!(prio, vec![0, 10, 5]);
    assert_eq!(pred, vec![usize::MAX, 0, 1]);
}
